workflow/workflow_test.go

```diff
@@ -1367,6 +1380,225 @@ func TestExpressionWithWhitespace(t *testing.T) {
+		t.Fatalf("incorrect error type returned: %T", err)
+	}
+	assert.LessThan(t, duration.Seconds(), 4) // It will take 5 seconds if it fails to fail early.
+}
+
+var multiDependencyFailureWorkflowWithErrorOut = `
+version: v0.2.0
+input:
+  root: WorkflowInput
+  objects:
+    WorkflowInput:
+      id: WorkflowInput
+      properties:
+        fail_purposefully:
+          type:
+            type_id: bool
+steps:
+  disabled_step:
+    plugin:
+      src: "n/a"
+      deployment_type: "builtin"
+    step: hello
+    input:
+      fail: !expr $.input.fail_purposefully
+  long_wait:
+    plugin:
+      src: "n/a"
+      deployment_type: "builtin"
+    step: wait
+    input:
+      wait_time_ms: 7000
+outputs:
+  workflow-success:
+    disabled_output: !expr $.steps.disabled_step.outputs.success
+    long_wait_output: !expr $.steps.long_wait.outputs.success
+`
+
+func TestMultiDependencyWorkflowFailureWithErrorOut(t *testing.T) {
+	// Tests failure when one dependency fails immediately, and the other one fails later.
+	preparedWorkflow := assert.NoErrorR[workflow.ExecutableWorkflow](t)(
+		getTestImplPreparedWorkflow(t, multiDependencyFailureWorkflowWithErrorOut),
+	)
+	startTime := time.Now() // Right before execute to not include pre-processing time.
+	_, _, err := preparedWorkflow.Execute(context.Background(), map[string]any{
+		"fail_purposefully": true,
+	})
+	duration := time.Since(startTime)
+	assert.Error(t, err)
+	t.Logf("MultiDependency workflow failed purposefully in %d ms", duration.Milliseconds())
+	var typedError *workflow.ErrNoMorePossibleOutputs
+	if !errors.As(err, &typedError) {
+		t.Fatalf("incorrect error type returned: %T", err)
+	}
+	// If it takes 5 seconds, then there was a deadlock in the client.
+	// If it takes 6 seconds, then it waited for the second step.
+	assert.LessThan(t, duration.Milliseconds(), 5500)
+}
+
+var multiDependencyFailureWorkflowWithDeployFailure = `
+version: v0.2.0
+input:
+  root: WorkflowInput
+  objects:
+    WorkflowInput:
+      id: WorkflowInput
+      properties:
+        fail_purposefully:
+          type:
+            type_id: bool
+steps:
+  failed_step:
+    plugin:
+      src: "n/a"
+      deployment_type: "builtin"
+    step: wait
+    input:
+      wait_time_ms: 0
+    deploy:
+      deployer_name: "test-impl"
+      deploy_succeed: !expr '!$.input.fail_purposefully'
+  long_wait:
+    plugin:
+      src: "n/a"
+      deployment_type: "builtin"
+    step: wait
+    input:
+      wait_time_ms: 7000
+outputs:
+  workflow-success:
+    simple_wait_output: !expr $.steps.failed_step.outputs
+    simple_wait_output: !expr $.steps.long_wait.outputs
+`
+
+func TestMultiDependencyWorkflowFailureWithDeployFailure(t *testing.T) {
+	// Tests failure when one dependency fails immediately, and the other one fails later.
+	// In this specific test the output depends on the `steps.step_name.outputs` node
+	// instead of the `steps.step_name.outputs.success` node because they are handled
```

Comment:
I'm not sure that using the generic "step_name" placeholders here helps anyone. It'd be just as easy to use the real step name, and easier to read.

Reply:
Changed.